### PR TITLE
Removing retry logic from HiveMetaStore to fix the metastore connection bloat

### LINKED ISSUE #54: Too many connections to the hive metastore
We observed that `kafka-connect-hdfs` creates too many connections to the hive metastore, essentially DoS-ing our servers. Ideally, there should never be more than `tasks.max` connections to the metastore since each `HdsSinkTask` creates one `DataWriter`, which creates one `HiveMetaStoreClient`. However, we observed thousands of connections when `tasks.max` was only 200. 

We believe the issue is caused by the retrying logic in `HiveMetaStore`. Specifally, `HiveMetaStore` blindly calls `reconnect` on all `TException`, without ever calling `client.close()`:

```
 74   private <R> R doWithRetry(ClientAction<R> action) throws TException {
 75     try {
 76       return action.call();
 77     } catch (TException e) {
 78       try {
 79           client.reconnect();
 80       } catch (MetaException swallowedException) {
 81         // reconnect failed, throw the original exception
 82         log.warn("Cannot connect to Hive meta store: {}", swallowedException.getMessage());
 83         throw e;
 84       }
 85       return action.call();
 86     }
 87   }

```

Moreover, adding a `client.close()` before `client.reconnect()` won't solve the problem as the underlying layers use a cacheable client wrapper that won't tear down the connection as long as there are users: https://github.com/apache/hive/blob/master/hcatalog/core/src/main/java/org/apache/hive/hcatalog/common/HiveClientCache.java#L394

**Resolution**: From what I understand, the retry logic is unnecessary since the underlying IMetaStoreClient object is actually the `RetryingHiveMetaStoreClient` (as opposed to the vanilla `HiveMetaStoreClient`) which already implements a sane retry logic (calls `reconnect` only after socket timeout). Therefore, we have solved the problem by simply removing the retry logic in `HiveMetaStore.java`

### PR DESCRIPTION
**Pull request by vikeshkhanna**
Fixes #54

### CODE REVIEW

**Comment by ewencp** [comment]
ok to test

**Comment by ewencp** [comment]
@Ishiihara Care to review?

**Comment by vikeshkhanna** [comment]
I am unable to fill the CLA form. It keeps timing out.

**Comment by vikeshkhanna** [comment]
Test failure seems like a build issue (ClassDefNotFoundError):
```
Caused by: java.lang.ClassNotFoundException: org.apache.commons.lang.StringUtils
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:259)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:235)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:227)
```

**Comment by ewencp** [comment]
@vikeshkhanna Sorry about the CLA issue, can you try again now?

**Comment by vikeshkhanna** [comment]
@ewencp Yes, works now. Submitted. Thanks!

**Comment by Ishiihara** [comment]
@ewencp Will take a look.

**Comment by Ishiihara** [comment]
@vikeshkhanna Thank you for the PR and the analysis of the issue. A couple of questions regarding this issue. First of all, is this change fix your problem? I think removing retries definitely reduce the number of connections as we make less number of retries, however, this may not solve the problem entirely. I think the root cause is that the connect tasks are long running and the cached MetaStoreClients are not removed from the cache and thus no connection will be shut down. As any retry will create a a new connection and no connection will be shutdown, eventually, this are still chances that the file descriptors are exhausted. 

One option to fix this issue though, is to set the `hcatalog.hive.client.cache.disabled` configuration in `hive.xml` so that we will not cache the client instance at all, so that any retry will first close the connection and then create a new one.

File: src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java [hunk 1]
```diff
@@ -71,19 +71,10 @@ public HiveMetaStore(Configuration conf, HdfsSinkConnectorConfig connectorConfig
...
-      } catch (MetaException swallowedException) {
-        // reconnect failed, throw the original exception
-        log.warn("Cannot connect to Hive meta store: {}", swallowedException.getMessage());
-        throw e;
-      }
-      return action.call();
-    }
+  private <R> R doAction(ClientAction<R> action) throws TException {
+    // No need to implement retries. We use RetryingHiveMetaStoreClient
+    // which already implements the retrying and reconnect logic
```
Comment (Ishiihara) [inline src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java]: Can you elaborate on what is the retry condition? It seems that two conditions can trigger retry: 1. The connection timeout is exceeded 2. We already retried and we should retry immediately. 

Also, do you know from which Hive version the `RetryingHiveMetaStoreClient` is included? Thanks!

Comment (vikeshkhanna) [inline src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java]: We hit the `AlreadExistsException` (which is a type of `TException`: https://hive.apache.org/javadocs/r0.10.0/api/org/apache/hadoop/hive/metastore/api/AlreadyExistsException.html) very often. This causes the `reconnect` to be called multiple times. Snippet from the logs of our forked code:

```
[2016-05-11 03:31:55,353] [pool-13-thread-1] **WARN [DEBBUG] Action failed. Reconnecting. (io.confluent.connect.hdfs.hive.HiveMetaStore:80)**
AlreadyExistsException(message:Partition already exists:Partition(values:[2016-05-10, 1462849200], dbName:xxxxx, tableName:client_timi
```

We package against the JARs specified in `kafka-connect-hdfs` pom.xml: So I'm guessing the Hive version is 1.2.1? Our deployed Hive version is also 1.2.1.

**Comment by vikeshkhanna** [comment]
@Ishiihara Thanks for looking into this issue!

Our initial investigation (stopped, started and ran Connect in Distributed mode about ~10 times) shows that the number of connections peaked at ~`tasks.max` after the fix. Earlier, at start-up time, the number of connections would go into thousands. I do not have a deep understanding of the caching client or the retrying client, but it seemed clear to me that the `RetryingMetaStoreClient` already implements the `reconnect` logic on failures and is also conscious of the socket timeout (does not call `reconnect` if the connection has been open for less than the timeout). Based solely on the latter, I think the higher level retry was unnecessary

File: src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java (hunk 1 above)
Comment (Ishiihara) [inline src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java]: That makes sense. I think it worthwhile to remove the retry in the HDFS connector's `HiveMetaStore`.  Would you like to update the comments to elaborate how the `RetryingHiveMetaStoreClient` handles retry?

**Comment by Ishiihara** [comment]
@vikeshkhanna Agreed that the high level retry seems unnecessary. The PR looks good in general and would be nice if you can try to see if setting `hcatalog.hive.client.cache.disabled` solves the issue. Also seems that the `RetryingMetaStoreClient` is added in Hive 0.10.0.

**Comment by vikeshkhanna** [comment]
@Ishiihara I'll try the setting once we resolve a few more blocking operational issues on our side :)

**Comment by Ishiihara** [comment]
@vikeshkhanna Mind to squash the commits? Then I will merge.

**Comment by Ishiihara** [comment]
retest this please

**Comment by vikeshkhanna** [comment]
@Ishiihara Squashed.

**Comment by Ishiihara** [comment]
retest this please

**Comment by Ishiihara** [comment]
LGTM. @vikeshkhanna Thanks for the contribution.

**Pull request merged**